exercises/src/main/scala/fpinscala/monads/Monad.scala

```diff
@@ -34,25 +34,49 @@ trait Monad[M[_]] extends Functor[M] {
   def map2[A,B,C](ma: M[A], mb: M[B])(f: (A, B) => C): M[C] =
     flatMap(ma)(a => map(mb)(b => f(a, b)))
 
-  def sequence[A](lma: List[M[A]]): M[List[A]] = ???
-
-  def traverse[A,B](la: List[A])(f: A => M[B]): M[List[B]] = ???
-
-  def replicateM[A](n: Int, ma: M[A]): M[List[A]] = ???
-
-  def compose[A,B,C](f: A => M[B], g: B => M[C]): A => M[C] = ???
+  def sequence[A](lma: List[M[A]]): M[List[A]] =
+    lma.foldRight(unit(Nil:List[A]))((p, acc) => {
+      map2(p, acc)(_ :: _)
+    })
+
+  def traverse[A,B](la: List[A])(f: A => M[B]): M[List[B]] =
+    la.foldRight(unit(Nil:List[B]))((p, acc) => {
+      map2(f(p), acc)(_ :: _)
+    })
+
+  def replicateM[A](n: Int, ma: M[A]): M[List[A]] =
+    if (n <= 0)
+      unit(Nil)
+    else
+      map2(ma, replicateM(n - 1, ma))(_ :: _)
+
+  def filterM[A](ms: List[A])(f: A => M[Boolean]): M[List[A]] =
+    ms.foldRight(unit(Nil:List[A]))((a, acc) => {
+      flatMap(f(a)) { p =>
+        if (p)
+          map2(unit(a), acc)(_ :: _)
+        else
+          acc
+      }
+    })
+
+  def compose[A,B,C](f: A => M[B], g: B => M[C]): A => M[C] =
+    a => flatMap(f(a))(b => g(b))
 
   // Implement in terms of `compose`:
-  def _flatMap[A,B](ma: M[A])(f: A => M[B]): M[B] = ???
+  def _flatMap[A,B](ma: M[A])(f: A => M[B]): M[B] =
+    compose((_: Int) => ma, f)(42)
```

Comment:
That `42` is there just as a dummy value that you don't care about if I understand correctly? Beside its appeal to Hitchhiker's Guide fans, using `()` of type `Unit` would probably be more clear.